ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -99,46 +102,54 @@ properties = emptyProperties
+codeLensProvider ideState pId CodeLensParams{_textDocument = TextDocumentIdentifier uri} = pluginResponse $ do
+    mode <- liftIO $ runAction "codeLens.config" ideState $ usePropertyAction #mode pId properties
+    nfp <- getNormalizedFilePath uri
+    env <- hscEnv . fst
+            <$> (handleMaybeM "Unable to get GhcSession"
+                $ liftIO
+                $ runAction "codeLens.GhcSession" ideState (useWithStale GhcSession nfp)
+                )
+    tmr <- fst <$> (
+                handleMaybeM "Unable to TypeCheck"
+              $ liftIO
+              $ runAction "codeLens.TypeCheck" ideState (useWithStale TypeCheck nfp)
+              )
+    bindings <- fst <$> (
+                handleMaybeM "Unable to GetBindings"
+                $ liftIO
+                $ runAction "codeLens.GetBindings" ideState (useWithStale GetBindings nfp)
+                )
+    (gblSigs@(GlobalBindingTypeSigsResult gblSigs'), gblSigsMp) <-
+      handleMaybeM "Unable to GetGlobalBindingTypeSigs"
+      $ liftIO
+      $ runAction "codeLens.GetGlobalBindingTypeSigs" ideState (useWithStale GetGlobalBindingTypeSigs nfp)
+
+    diag <- liftIO $ atomically $ getDiagnostics ideState
+    hDiag <- liftIO $ atomically $ getHiddenDiagnostics ideState
+
+    let toWorkSpaceEdit tedit = WorkspaceEdit (Just $ Map.singleton uri $ List tedit) Nothing Nothing
+        generateLensForGlobal mp sig@GlobalBindingTypeSig{gbRendered} = do
+            range <- toCurrentRange mp =<< srcSpanToRange (gbSrcSpan sig)
+            tedit <- gblBindingTypeSigToEdit sig (Just gblSigsMp)
             let wedit = toWorkSpaceEdit [tedit]
             pure $ generateLens pId range (T.pack gbRendered) wedit
-          gblSigs' = maybe [] (\(GlobalBindingTypeSigsResult x) -> x) gblSigs
-          generateLensFromDiags f =
-            sequence
-              [ pure $ generateLens pId _range title edit
+        generateLensFromDiags f =
+              [ generateLens pId _range title edit
               | (dFile, _, dDiag@Diagnostic{_range = _range}) <- diag ++ hDiag
-              , dFile == filePath
+              , dFile == nfp
               , (title, tedit) <- f dDiag
               , let edit = toWorkSpaceEdit tedit
               ]
-
-      case mode of
+    pure $ List $ case mode of
         Always ->
-          pure (catMaybes $ generateLensForGlobal <$> gblSigs')
-            <> generateLensFromDiags (suggestLocalSignature False env tmr bindings) -- we still need diagnostics for local bindings
-        Exported -> pure $ catMaybes $ generateLensForGlobal <$> filter gbExported gblSigs'
-        Diagnostics -> generateLensFromDiags $ suggestSignature False env gblSigs tmr bindings
-    Nothing -> pure []
+          mapMaybe (generateLensForGlobal gblSigsMp) gblSigs'
+            <> generateLensFromDiags
+                (suggestLocalSignature False (Just env) (Just tmr) (Just bindings)) -- we still need diagnostics for local bindings
```

Comment:
Here it looks like after generating stuff from the diagnostics, we're going to apply this function to the diagnostics. But this function takes `bindings`, which might need position mapping. Maybe I'm totally wrong here, it just seems suspicious.

Reply:
`suggestLocalSignature` relies on diagnostic, if diagnostics don't have the local signature warning, the `bindings` is useless, and if diagnostic has, that means we parsed success, and the `bindings` is fresh.
https://github.com/haskell/haskell-language-server/blob/6f5a73507f8d9266a486feaf8695c052362b9b95/ghcide/src/Development/IDE/Plugin/TypeLenses.hs#L171-L176

Reply:
Okay. Worth a comment!

Reply:
Done